### ISSUE TITLE
Incorrectly narrowed Literal in while loops

### ISSUE THREAD

**Issue by osandov** [python/mypy#17096]
**Bug Report**

The following code snippet:

```python
from typing import Literal


class Node:
    color: Literal["red", "black"]
    next: Node


def walk(node: Node) -> None:
    while node.color == "red":
        node = node.next
    reveal_type(node.color)
    while True:
        reveal_type(node.color)
        if node.color == "red":
            return
        node = node.next
        reveal_type(node.color)
```

Results in a `comparison-overlap` false positive under `mypy --strict`:

```
test.py:12: note: Revealed type is "Literal['black']"
test.py:14: note: Revealed type is "Literal['black']"
test.py:15: error: Non-overlapping equality check (left operand type: "Literal['black']", right operand type: "Literal['red']")  [comparison-overlap]
test.py:18: note: Revealed type is "Union[Literal['red'], Literal['black']]"
Found 1 error in 1 file (checked 1 source file)
```

As you can see, mypy correctly concludes that at line 12, `node.color` must be `"black"`, and at line 18, it may be `"red"` or `"black"`. But at line 14, it still thinks that it must be `"black"` even though it may not be on subsequent iterations. This causes the false positive on line 15. 

**To Reproduce**

https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=3207b91810da9f447ff090d86df689a5

**Your Environment**

- Mypy version used: tested on 1.10.0+dev.80190101f68b52e960c22572ed6cc814de078b9c and 1.8.0
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.12.2

**Comment by Udit19-pixel** [comment]
Can you tell me where you found this snippet?
I can try helping after knowing this :)

**Comment by osandov** [comment]
The real code where I hit this is more complicated, and the affected variable is an `enum.Enum` instead of a `typing.Union`, but it appears to be the same issue as my minimal example: https://github.com/osandov/drgn/blob/8ab4901d8336faf74bdd5f0b6bf8127d87d818f9/drgn/helpers/common/type.py#L176

**Comment by Udit19-pixel** [comment]
Thank you for providing more context by giving out a reference.
Upon closer look to the member_at_offset function, the variable type. kind can be mapped with node.color inside the while loop.

At the start of the while loop, mypy would think that color is the same as start_color. Then first reveal_type would check only for black and red literal, but the second reveal_type might check for both. So, there will be a type-inference checking after every possible modifications, which can be updated and the unionType can be carried forward.

I understand that implementing this solution is complex. If the example reference could assist you in developing the actual code, I would be grateful as I am a novice in open source contributions and I am still trying to understand how the code-base is implemented. 

Following is the reference solution which I think will be able to solve -

```from typing import Literal, Union
def color_walk(start_color: Literal["red", "black"]) -> None:
    color: Literal["red", "black"] = start_color
    while True:
        reveal_type(color)
        
        if color == "black":
            color = "red"
        else:
            color = "black"
     
        reveal_type(color)
        
        if sample_condition_block(): 
            break```

**Comment by agronholm** [comment]
I hit this issue too:
```py
from enum import Enum


class MyEnum(Enum):
    value1 = 1
    value2 = 2


class Foo:
    def __init__(self) -> None:
        self.state = MyEnum.value1

    def change(self) -> None:
        self.state = MyEnum.value2

foo = Foo()
assert foo.state is MyEnum.value1
foo.change()
assert foo.state is MyEnum.value2
```
Resulted in:
```
mypytest.py:19: error: Non-overlapping identity check (left operand type: "Literal[MyEnum.value1]", right operand type: "Literal[MyEnum.value2]")  [comparison-overlap]
Found 1 error in 1 file (checked 1 source file)
```
It should not assume that, if the value was a specific literal item now, it will stay that way forever (unless it was annotated with `Final`).
